keystore/v2/keystore/signature/notary.go

```diff
@@ -103,11 +110,11 @@ func (s *Notary) Verify(data, context []byte) (*asn1.VerifiedContainer, error) {
 }
 
 func (s *Notary) signData(data, context []byte) []asn1.Signature {
-	signatures := make([]asn1.Signature, 0, 1)
-	signatures = append(signatures, asn1.Signature{
-		Algorithm: asn1.Sha256OID,
-		Signature: s.signSHA256(data, context),
-	})
+	signatures := make([]asn1.Signature, len(s.algorithms))
+	for i, algo := range s.algorithms {
```

Comment:
can you recall when and for what we decided to use several signatures with different algos for same data?

Reply:
> can you recall when and for what we decided to use several signatures with different algos for same data?

It’s mostly for future-proofing.

Indeed, in the nearest foreseeable future we are going to use HMAC-SHA-256 exclusively. Supporting more than one possible algorithm allows us to add new algorithms and phase out old ones without it being a breaking change requiring a flag day when upgrading Acra installation.

With this we can simply add a new algorithm and it will be gradually used after update. We can remove an outdated algorithm and existing signatures will simply be ignored. We do not have to develop any conversion utilities for resigning keystore with new algorithm.

But yeah, there is no immediate short-term benefit in it. This comes into play when SHA-256 will be declared not strong enough anymore.

Reply:
okay, to avoid breaking changes and re-signing logs, we can try to verify with all supported algorithms. but what the benefit from signing with different algos instead just to use last/new added?

Reply:
> but what the benefit from signing with different algos instead just to use last/new added?

It somewhat increases resistance to collisions. Even if individual algorithms are easy to crack, it’s much harder to find a value which results in a simultaneous collision of multiple hash functions based on different computation principles. For example, if we add BLAKE3 along with SHA-256, then even if SHA-256 is cracked tomorrow, user key stores will be more or less shielded from that because existing key rings will have BLAKE3 signatures on them, which is radically different algorithm supposedly immune to collisions in SHA-2.

There isn’t much benefit in this if we update the current signature algorithm in a timely manner and these updates are distributed and applied to production environment in a timely manner. However, it’s easier said than done. Supporting multiple algorithms allows us to do it smoother and ahead of time.